Fancyauth.APIUtil/FancyPluginBase.cs

```diff
@@ -0,0 +1,363 @@
+
+        public async sealed override Task OnUserConnected(IUser user)
+        {
+            await OnThingAdded(GlobalUserListeners, SingleUserListeners, UserListeners, user, user.SessionId);
+        }
+
+        public async override Task OnChannelCreated(IChannel channel)
+        {
+            await OnThingAdded(GlobalChannelListeners, SingleChannelListeners, ChannelListeners, channel, channel.ChannelId);
+        }
+
+        private async Task OnThingAdded<T>(ConcurrentBag<MuTuple<Func<T, object>, Func<T, Task>>> globalListeners,
+                ConcurrentDictionary<int, IList<MuTuple<Func<T, object>, Func<T, Task>>>> singleListeners,
+                ConcurrentDictionary<int, IList<MuTuple<Func<T, object>, Func<T, Task>, object>>> listeners,
+                T thing, int uniqueId) where T : IThing, IReadModifyWriteObject
+        {
+            // whenever a thing (e.g. user) is added, all registered listeners for this thing must be added
+            // first get all registered users by getting all global listeners and registered single listeners
+            foreach (var t in globalListeners.Concat(singleListeners[uniqueId])) {
+                // create the tuple to add (mapper, listener, mapper(thing))
+                var tuple = MuTuple.Create(t.Item1, t.Item2, t.Item1(thing));
+                // add each of them to the listeners list
+                // if it is the first one, a new List must be created and the object be added
+                // otherwise add it to the list
+                listeners.AddOrUpdate(uniqueId,
+                    // add new list
+                    i =>
+                    {
+                        var l = new List<MuTuple<Func<T, object>, Func<T, Task>, object>>();
+                        l.Add(tuple);
+                        return l;
+                    },
+                    // update list
+                    (i, l) =>
+                    {
+                        l.Add(tuple);
+                        return l;
+                    }
+                );
+            }
+
+            // execute all listeners async
+            await Task.WhenAll(listeners[uniqueId].Select(t => t.Item2(thing)).AsEnumerable());
+            // save changes
+            await thing.SaveChanges();
+        }
+
+        public async sealed override Task OnUserDisconnected(IUser user)
+        {
+            await OnThingRemoved(UserListeners, user.SessionId);
+        }
+
+        public async override Task OnChannelDeleted(IChannel channel)
+        {
+            await OnThingRemoved(ChannelListeners, channel.ChannelId);
+        }
+
+        private async Task OnThingRemoved<T>(ConcurrentDictionary<int, IList<MuTuple<Func<T, object>, Func<T, Task>, object>>> listeners, int uniqueId) where T : IThing
+        {
+            IList<MuTuple<Func<T, object>, Func<T, Task>, object>> dontCare;
+            listeners.TryRemove(uniqueId, out dontCare);
+            await Task.FromResult((object)null);
```

Comment:
No need to await, you can just return this (and remove the `async` in the signature)